source/fuzz/transformation_propagate_instruction_up.cpp

```diff
@@ -0,0 +1,393 @@
+         std::all_of(maybe_fresh_ids.begin(), maybe_fresh_ids.end(),
+                     [ir_context](uint32_t id) {
+                       return fuzzerutil::IsFreshId(ir_context, id);
+                     });
+}
+
+void TransformationPropagateInstructionUp::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto* inst = GetInstructionToPropagate(ir_context, message_.block_id());
+  assert(inst &&
+         "The block must have at least one supported instruction to propagate");
+
+  // |inst| might depend on OpPhi instructions from the same basic block.
+  // |op_phi_to_result_id| contains a mapping from the result id of such an
+  // OpPhi instruction to the map of its operands
+  // (i.e. |op_phi_to_result_id[op_phi_id][label_id] == result_id|).
+  auto op_phi_to_result_id =
+      ComputeMappingFromOpPhiToResultId(ir_context, inst);
+
+  opt::Instruction::OperandList op_phi_operands;
+  for (auto predecessor_id : ir_context->cfg()->preds(message_.block_id())) {
+    auto new_result_id =
+        message_.predecessor_id_to_fresh_id().at(predecessor_id);
+
+    // Compute InOperands for the OpPhi instruction to be inserted later.
+    op_phi_operands.push_back({SPV_OPERAND_TYPE_ID, {new_result_id}});
+    op_phi_operands.push_back({SPV_OPERAND_TYPE_ID, {predecessor_id}});
+
+    // Create a clone of the |inst| to be inserted into the |predecessor_id|.
+    std::unique_ptr<opt::Instruction> clone(inst->Clone(ir_context));
+    clone->SetResultId(new_result_id);
+
+    fuzzerutil::UpdateModuleIdBound(ir_context, new_result_id);
+
+    // Adjust |clone|'s operands to account for possible dependencies on OpPhi
+    // instructions from the same basic block.
+    for (uint32_t i = 0; i < clone->NumInOperands(); ++i) {
+      auto& operand = clone->GetInOperand(i);
+
+      if (operand.type == SPV_OPERAND_TYPE_ID &&
+          op_phi_to_result_id.count(operand.words[0])) {
+        operand.words[0] =
+            op_phi_to_result_id.at(operand.words[0]).at(predecessor_id);
+      }
+    }
+
+    // Insert cloned instruction into the predecessor.
+    auto* predecessor = ir_context->cfg()->block(predecessor_id);
+    assert(predecessor && "|predecessor_id| is invalid");
+
+    GetLastInsertBeforeInstruction(predecessor, clone->opcode())
+        ->InsertBefore(std::move(clone));
+  }
+
+  // Insert an OpPhi instruction into the basic block of |inst|.
+  ir_context->get_instr_block(inst)->begin()->InsertBefore(
+      MakeUnique<opt::Instruction>(ir_context, SpvOpPhi, inst->type_id(),
+                                   inst->result_id(),
+                                   std::move(op_phi_operands)));
+
+  // Remove |inst| from the basic block.
+  inst->RemoveFromList();
+  delete inst;  // RemoveFromList doesn't clear the memory.
+
+  // Make sure our changes are analyzed
+  ir_context->InvalidateAnalysesExceptFor(
+      opt::IRContext::Analysis::kAnalysisNone);
+
+  // Remove unused OpPhi instructions.
```

Comment:
I would explain this in a bit more detail: it's possible that the instruction that has been propagated might have been the only user of an OpPhi instruction, in which case we can delete that instruction.

Reply:
Fixed in c74f0e3.